src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs

```diff
@@ -85,14 +90,26 @@ await _renderer.InitializeStandardComponentServicesAsync(
         await using var writer = new HttpResponseStreamWriter(context.Response.Body, Encoding.UTF8, defaultBufferSize, ArrayPool<byte>.Shared, ArrayPool<char>.Shared);
         using var bufferWriter = new BufferedTextWriter(writer);
 
+        int originalStatusCode = context.Response.StatusCode;
+
         // Note that we always use Static rendering mode for the top-level output from a RazorComponentResult,
         // because you never want to serialize the invocation of RazorComponentResultHost. Instead, that host
         // component takes care of switching into your desired render mode when it produces its own output.
         var htmlContent = await _renderer.RenderEndpointComponent(
             context,
             rootComponent,
             ParameterView.Empty,
-            waitForQuiescence: result.IsPost || isErrorHandler);
+            waitForQuiescence: result.IsPost || isErrorHandler || hasStatusCodePage);
+
+        bool requresReexecution = originalStatusCode != context.Response.StatusCode && hasStatusCodePage;
+        if (requresReexecution)
+        {
+            // If the response is a 404, we don't want to write any content.
+            // This is because the 404 status code is used by the routing middleware
+            // to indicate that no endpoint was found for the request.
+            await bufferWriter.FlushAsync();
+            return;
+        }
```

Comment:
Requires re-execution
```suggestion
        bool requiresReexecution = originalStatusCode != context.Response.StatusCode && hasStatusCodePage;
        if (requiresReexecution)
        {
            // If the response is a 404, we don't want to write any content.
            // This is because the 404 status code is used by the routing middleware
            // to indicate that no endpoint was found for the request.
            await bufferWriter.FlushAsync();
            return;
        }
```